File: src/services/AppointmentService.ts
import { Appointment } from '../models/Appointment';
import { Reminder, ReminderMethod } from '../models/Reminder';
import { AppointmentRepository } from '../repositories/AppointmentRepository';
import { GroupMeetingRepository } from '../repositories/GroupMeetingRepository';
import { AddAppointmentRequest } from '../models/types';

export class AppointmentService {
  constructor(
    private appointmentRepo: AppointmentRepository,
    private groupMeetingRepo: GroupMeetingRepository,
  ) {}

  async getUserAppointments(userId: number): Promise<Appointment[]> {
    return this.appointmentRepo.getByUserId(userId);
  }

  async getDetails(appointmentId: number): Promise<Appointment | null> {
    return this.appointmentRepo.getById(appointmentId);
  }

  async checkPersonalConflict(
    userId: number,
    start: Date,
    end: Date,
  ): Promise<Appointment | null> {
    return this.appointmentRepo.findConflicts(userId, start, end);
  }

  async createPersonal(userId: number, request: AddAppointmentRequest): Promise<number> {
    const newApp = new Appointment(
      0,
      request.title.trim(),
      request.location,
      request.startTime,
      request.endTime,
      userId,
      false,
    );

    this.addSelectedReminders(newApp, request.reminderMethods);

    return this.appointmentRepo.save(newApp);
  }

  async createGroup(userId: number, request: AddAppointmentRequest): Promise<number> {
    const newApp = new Appointment(
      0,
      request.title.trim(),
      request.location,
      request.startTime,
      request.endTime,
      userId,
      true,
    );

    this.addSelectedReminders(newApp, request.reminderMethods);

    const appId = await this.appointmentRepo.save(newApp);

    await this.groupMeetingRepo.saveGroupMetadata(appId);
    await this.groupMeetingRepo.addParticipant(userId, appId);

    return appId;
  }

  async replaceAppointment(
    oldId: number,
    userId: number,
    request: AddAppointmentRequest,
  ): Promise<number> {
    await this.appointmentRepo.delete(oldId);

    if (request.isGroupMeeting) {
      return this.createGroup(userId, request);
    }

    return this.createPersonal(userId, request);
  }

  private addSelectedReminders(
    appointment: Appointment,
    methods?: ReminderMethod[],
  ): void {
    if (!methods || methods.length === 0) return;

    for (const method of methods) {
      const remindAt = new Date(appointment.startTime);
      remindAt.setMinutes(remindAt.getMinutes() - 15);

      appointment.addReminder(new Reminder(0, remindAt, method));
    }
  }
}

File: CalendarApi/Services/AppointmentService.cs
using CalendarApi.DTOs;
using CalendarApi.Models;
using CalendarApi.Repositories;

namespace CalendarApi.Services;

/// <summary>
/// Business logic for Appointment operations.
/// Migrated from TypeScript AppointmentService.ts
/// </summary>
public class AppointmentService
{
    private readonly AppointmentRepository _appointmentRepo;
    private readonly GroupMeetingRepository _groupMeetingRepo;

    public AppointmentService(AppointmentRepository appointmentRepo, GroupMeetingRepository groupMeetingRepo)
    {
        _appointmentRepo = appointmentRepo;
        _groupMeetingRepo = groupMeetingRepo;
    }

    public List<Appointment> GetUserAppointments(int userId)
    {
        return _appointmentRepo.GetByUserId(userId);
    }

    public Appointment? GetDetails(int appointmentId)
    {
        return _appointmentRepo.GetById(appointmentId);
    }

    public Appointment? CheckPersonalConflict(int userId, DateTime start, DateTime end)
    {
        return _appointmentRepo.FindConflicts(userId, start, end);
    }

    public int CreatePersonal(int userId, AddAppointmentRequest request)
    {
        var newApp = new Appointment(
            0,
            request.Title.Trim(),
            request.Location,
            request.StartTime,
            request.EndTime,
            userId,
            isGroupMeeting: false
        );

        AddSelectedReminders(newApp, request.ReminderMethods);
        return _appointmentRepo.Save(newApp);
    }

    public int CreateGroup(int userId, AddAppointmentRequest request)
    {
        var newApp = new Appointment(
            0,
            request.Title.Trim(),
            request.Location,
            request.StartTime,
            request.EndTime,
            userId,
            isGroupMeeting: true
        );

        AddSelectedReminders(newApp, request.ReminderMethods);

        var appId = _appointmentRepo.Save(newApp);
        _groupMeetingRepo.SaveGroupMetadata(appId);
        _groupMeetingRepo.AddParticipant(userId, appId);

        return appId;
    }

    public int ReplaceAppointment(int oldId, int userId, AddAppointmentRequest request)
    {
        _appointmentRepo.Delete(oldId);

        if (request.IsGroupMeeting)
            return CreateGroup(userId, request);

        return CreatePersonal(userId, request);
    }

    public void DeleteAppointment(int appointmentId)
    {
        _appointmentRepo.Delete(appointmentId);
    }

    private static void AddSelectedReminders(Appointment appointment, List<ReminderMethod>? methods)
    {
        if (methods == null || methods.Count == 0) return;

        foreach (var method in methods)
        {
            var remindAt = appointment.StartTime.AddMinutes(-15);
            appointment.AddReminder(new Reminder(0, remindAt, method));
        }
    }
}
